lens-cube/src/main/java/org/apache/lens/cube/parse/ExpressionResolver.java

```diff
@@ -647,37 +647,36 @@ public void rewriteContext(CubeQueryContext cubeql) throws LensException {
       // prune invalid expressions
       cubeql.getExprCtx().pruneExpressions();
       // prune candidate facts without any valid expressions
-      if (cubeql.getCube() != null && !cubeql.getCandidateFacts().isEmpty()) {
+      if (cubeql.getCube() != null && !cubeql.getCandidates().isEmpty()) {
         for (Map.Entry<String, Set<ExpressionContext>> ecEntry : exprCtx.allExprsQueried.entrySet()) {
           String expr = ecEntry.getKey();
           Set<ExpressionContext> ecSet = ecEntry.getValue();
           for (ExpressionContext ec : ecSet) {
             if (ec.getSrcTable().getName().equals(cubeql.getCube().getName())) {
               if (cubeql.getQueriedExprsWithMeasures().contains(expr)) {
-                for (Iterator<Set<CandidateFact>> sItr = cubeql.getCandidateFactSets().iterator(); sItr.hasNext();) {
-                  Set<CandidateFact> factSet = sItr.next();
+                for (Iterator<Candidate> sItr = cubeql.getCandidates().iterator(); sItr.hasNext();) {
+                  Candidate cand = sItr.next();
                   boolean evaluableInSet = false;
-                  for (CandidateFact cfact : factSet) {
-                    if (ec.isEvaluable(cfact)) {
-                      //TODO union: in case of join , one of the candidates should be able to answer the mesaure expression
-                      //TODO union: In case of union, all the candidates should answer the expression
-                      //TODO union : add isExpresionEvaluable() to Candidate
-                      evaluableInSet = true;
-                    }
+                  if (cand.isExpressionEvaluable(ec)) {
+                    evaluableInSet = true;
                   }
                   if (!evaluableInSet) {
-                    log.info("Not considering fact table set:{} as {} is not evaluable", factSet, ec.exprCol.getName());
+                    log.info("Not considering Candidate :{} as {} is not evaluable", cand, ec.exprCol.getName());
                     sItr.remove();
                   }
                 }
               } else {
-                for (Iterator<CandidateFact> i = cubeql.getCandidateFacts().iterator(); i.hasNext();) {
-                  CandidateFact cfact = i.next();
-                  if (!ec.isEvaluable(cfact)) {
-                    log.info("Not considering fact table:{} as {} is not evaluable", cfact, ec.exprCol.getName());
-                    cubeql.addFactPruningMsgs(cfact.fact,
-                      CandidateTablePruneCause.expressionNotEvaluable(ec.exprCol.getName()));
-                    i.remove();
+                for (Iterator<Candidate> i = cubeql.getCandidates().iterator(); i.hasNext();) {
+                  Candidate cand = i.next();
+                  Set<StorageCandidate> scSet = CandidateUtil.getStorageCandidates(cand);
+                  for (StorageCandidate sc : scSet) {
+                    if (!sc.isExpressionEvaluable(ec)) {
+                      i.remove();
```

Comment:
This will not work since in outer look we are iterating over candidates and  this call will modify the candidates set. 

Can we have a separate for loop for all Dim Candidates.